feat: add C implementation for `math/base/special/heaviside`

sahil20021008 opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: passed
  - task: lint_repl_help status: na
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: na
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: passed
  - task: lint_c_examples status: passed
  - task: lint_c_benchmarks status: passed
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: na
  - task: lint_typescript_tests status: passed
  - task: lint_license_headers status: passed ---

Resolves #1891 .

## Description

> What is the purpose of this pull request?

This pull request:

-   proposes adding C implementation of `stdlib/math/base/special/heaviside`. This includes benchmarks, examples, tests, README files, etc.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #649 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

sahil20021008 committed:
feat: add c implementation of stdlib/math/base/special/heaviside

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

sahil20021008 committed:
feat: add implementation of stdlib/math/base/special/heaviside

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

sahil20021008 committed:
feat: add implementation of stdlib/math/base/special/heaviside

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/heaviside/README.md:
```diff
@@ -2,7 +2,7 @@
 @license Apache-2.0
 
-Copyright (c) 2018 The Stdlib Authors.
+Copyright (c) 2025 The Stdlib Authors.
```
```suggestion
Copyright (c) 2018 The Stdlib Authors.
```

gunjjoshi replied:
We can avoid updating copyright years for pre-existing files.

sahil20021008 replied:
Done

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/heaviside/benchmark/c/Makefile:
```diff
@@ -1,7 +1,7 @@
 # @license Apache-2.0
 #
-# Copyright (c) 2018 The Stdlib Authors.
+# Copyright (c) 2025 The Stdlib Authors.
```
```suggestion
# Copyright (c) 2018 The Stdlib Authors.
```

sahil20021008 replied:
Done

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/heaviside/benchmark/c/benchmark.c:
```diff
@@ -1,7 +1,7 @@
 * @license Apache-2.0
 *
-* Copyright (c) 2018 The Stdlib Authors.
+* Copyright (c) 2025 The Stdlib Authors.
```
```suggestion
* Copyright (c) 2018 The Stdlib Authors.
```

sahil20021008 replied:
Done

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/heaviside/docs/types/test.ts:
```diff
@@ -23,7 +23,7 @@ import heaviside = require( './index' );
 // The function returns a number...
 {
-	heaviside( 3.141592653589793 ); // $ExpectType number
+	heaviside( 3.141592653589793, 'half-maximum' ); // $ExpectType number
```
I don't think this is necessary, @sahil20021008. We want to demonstrate the default case here, without the second argument, such as the function is discontinuous at `0`.

gunjjoshi replied:
You can revert this change.

sahil20021008 replied:
Done

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/heaviside/src/main.c:
```diff
@@ -0,0 +1,76 @@
+*
+* @param x 				input value
+* @param continuity 	continuity option
+* @return				function value
```
The indentation is off here.

sahil20021008 replied:
Done

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/heaviside/README.md:
```diff
@@ -164,6 +164,105 @@ logEachMap( 'H(%0.4f) = %0.4f', x, heaviside );
+
+#### stdlib_base_heaviside( x )
+
+Evaluate the [Heaviside function][heaviside-function].
```
```suggestion
Evaluates the [Heaviside function][heaviside-function].
```

sahil20021008 replied:
Done

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/heaviside/README.md:
````diff
@@ -164,6 +164,105 @@ logEachMap( 'H(%0.4f) = %0.4f', x, heaviside );
+#include "stdlib/math/base/special/heaviside.h"
+```
+
+#### stdlib_base_heaviside( x )
````
```suggestion
#### stdlib_base_heaviside( x, continuity )
```

sahil20021008 replied:
Done

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/heaviside/src/main.c:
```diff
@@ -0,0 +1,76 @@
+			return 0.5;
+		}
+		// Default behaviour is discontinuity
+		return 0.0/0.0;
```
Spacing is off at certain places here, will need to be fixed.

sahil20021008 replied:
Not exactly sure where the placing has to be improved
For now, I have fixed placing at line 58, but please lmk if changes required anywhere else also.
Thanks

sahil20021008 committed:
feat: add c implementation of heaviside

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

sahil20021008 reviewed (commented):
Hi @gunjjoshi 
I have addressed your review comments.

sahil20021008 commented:
Receiving the following error while using `git push`. Not sure why this is happening
```
Compiling benchmark...
make[1]: Entering directory '/workspaces/stdlib/lib/node_modules/@stdlib/math/base/special/heaviside/benchmark/c'
benchmark.c:24:10: fatal error: stdlib/math/base/special/heaviside.h: No such file or directory
   24 | #include "stdlib/math/base/special/heaviside.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[1]: *** [Makefile:105: benchmark.out] Error 1
make[1]: Leaving directory '/workspaces/stdlib/lib/node_modules/@stdlib/math/base/special/heaviside/benchmark/c'
Error when attempting to compile benchmark.
ERROR: An error was encountered during execution.

make: *** [/workspaces/stdlib/tools/make/lib/benchmark/c.mk:84: benchmark-c-files] Error 1

Encountered an error when running C benchmarks.
```
have used `git push --no-verify` for now.
Any idea what is the issue here @gunjjoshi ?
I was not facing this error earlier

sahil20021008 committed:
feat: add implementation of stdlib/math/base/special/heaviside

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented:
Ref: https://github.com/stdlib-js/stdlib/pull/10196
